File: RepoTransBench/target_projects/Python/Go/VerbalExpressions_PythonVerbalExpressions/tests/original/verbal_expressions_test.go
package original

import (
	"regexp"
	"testing"

	"verbalexpressions"
)

func TestShouldRenderVerExAsString(t *testing.T) {
	v := verbalexpressions.NewVerEx()
	v2 := v.Add("^$")
	if str := v2.String(); str != "^$" {
		t.Errorf("Expected '^$', got '%s'", str)
	}
}

func TestShouldRenderVerExListAsString(t *testing.T) {
	v := verbalexpressions.NewVerEx()
	v2 := v.Add([]string{"^", "[0-9]", "$"})
	if str := v2.String(); str != "^[0-9]$" {
		t.Errorf("Expected '^[0-9]$', got '%s'", str)
	}
}

func TestShouldMatchCharactersInRange(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Range("a", "c").Regex()
	for _, character := range []string{"a", "b", "c"} {
		if !exp.MatchString(character) {
			t.Errorf("Expected regex to match '%s'", character)
		}
	}
}

func TestShouldNotMatchCharactersOutsideOfRange(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Range("a", "c").Regex()
	if exp.MatchString("d") {
		t.Errorf("Expected regex to NOT match 'd'")
	}
}

func TestShouldMatchCharactersInExtendedRange(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Range("a", "b", "X", "Z").Regex()
	for _, character := range []string{"a", "b", "X", "Y", "Z"} {
		if !exp.MatchString(character) {
			t.Errorf("Expected to match '%s'", character)
		}
	}
}

func TestShouldNotMatchCharactersOutsideOfExtendedRange(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Range("a", "b", "X", "Z").Regex()
	for _, character := range []string{"c", "W"} {
		if exp.MatchString(character) {
			t.Errorf("Expected to NOT match '%s'", character)
		}
	}
}

func TestShouldMatchStartOfLine(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Regex()
	if !exp.MatchString("text  ") {
		t.Errorf("Expected regex to match 'text  ' (start-of-line)")
	}
}

func TestShouldMatchEndOfLine(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().EndOfLine().Regex()
	if !exp.MatchString("") {
		t.Errorf("Expected regex to match empty string at start and end-of-line")
	}
}

func TestShouldMatchAnything(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().EndOfLine().Regex()
	if !exp.MatchString("!@#$%¨&*()__+{}") {
		t.Errorf("Expected to match '!@#$%¨&*()__+{}'")
	}
}

func TestShouldMatchAnythingButSpecifiedElementWhenElementNotFound(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().AnythingBut("X").EndOfLine().Regex()
	if !exp.MatchString("Y Files") {
		t.Errorf("Expected to match 'Y Files' when 'X' is disallowed")
	}
}

func TestShouldNotMatchAnythingButSpecifiedElementWhenElementIsFound(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().AnythingBut("X").EndOfLine().Regex()
	if exp.MatchString("VerEX") {
		t.Errorf("Expected NOT to match 'VerEX' when 'X' is forbidden")
	}
}

func TestShouldFindElement(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Find("Wally").EndOfLine().Regex()
	if !exp.MatchString("Wally") {
		t.Errorf("Expected to match 'Wally'")
	}
}

func TestShouldNotFindMissingElement(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Find("Wally").EndOfLine().Regex()
	if exp.MatchString("Wall-e") {
		t.Errorf("Expected NOT to match 'Wall-e'")
	}
}

func TestShouldMatchWhenMaybeElementIsPresent(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Find("Python2.").Maybe("7").EndOfLine().Regex()
	if !exp.MatchString("Python2.7") {
		t.Errorf("Expected to match 'Python2.7'")
	}
}

func TestShouldMatchWhenMaybeElementIsMissing(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Find("Python2.").Maybe("7").EndOfLine().Regex()
	if !exp.MatchString("Python2.") {
		t.Errorf("Expected to match 'Python2.'")
	}
}

func TestShouldMatchOnAnyWhenElementIsFound(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Any("Q").Anything().EndOfLine().Regex()
	if !exp.MatchString("Query") {
		t.Errorf("Expected to match 'Query'")
	}
}

func TestShouldNotMatchOnAnyWhenElementIsNotFound(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Any("Q").Anything().EndOfLine().Regex()
	if exp.MatchString("W") {
		t.Errorf("Expected NOT to match 'W'")
	}
}

func TestShouldMatchWhenLineBreakPresent(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().LineBreak().Anything().EndOfLine().Regex()
	if !exp.MatchString("Marco \n Polo") {
		t.Errorf("Expected to match 'Marco \\n Polo'")
	}
}

func TestShouldMatchWhenLineBreakAndCarriageReturnPresent(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().LineBreak().Anything().EndOfLine().Regex()
	if !exp.MatchString("Marco \r\n Polo") {
		t.Errorf("Expected to match 'Marco \\r\\n Polo'")
	}
}

func TestShouldNotMatchWhenLineBreakIsMissing(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().LineBreak().Anything().EndOfLine().Regex()
	if exp.MatchString("Marco Polo") {
		t.Errorf("Expected NOT to match 'Marco Polo'")
	}
}

func TestShouldMatchWhenTabPresent(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().Tab().EndOfLine().Regex()
	if !exp.MatchString("One tab only\t") {
		t.Errorf("Expected to match 'One tab only\\t'")
	}
}

func TestShouldNotMatchWhenTabIsMissing(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().Tab().EndOfLine().Regex()
	if exp.MatchString("No tab here") {
		t.Errorf("Expected NOT to match 'No tab here'")
	}
}

func TestShouldMatchWhenWordPresent(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().Word().EndOfLine().Regex()
	if !exp.MatchString("Oneword") {
		t.Errorf("Expected to match 'Oneword'")
	}
}

func TestNotMatchWhenTwoWordsArePresentInsteadOfOne(t *testing.T) {
	// This test is checking that tab is not present between two words, so shouldn't match
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().Tab().EndOfLine().Regex()
	if exp.MatchString("Two words") {
		t.Errorf("Expected NOT to match 'Two words'")
	}
}

func TestShouldMatchWhenOrConditionFulfilled(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().Find("G").Or().Find("h").EndOfLine().Regex()
	if !exp.MatchString("Github") {
		t.Errorf("Expected to match 'Github'")
	}
}

func TestShouldNotMatchWhenOrConditionNotFulfilled(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().Find("G").Or().Find("h").EndOfLine().Regex()
	if exp.MatchString("Bitbucket") {
		t.Errorf("Expected NOT to match 'Bitbucket'")
	}
}

func TestShouldMatchOnUpperCaseWhenLowerCaseIsGivenAndAnyCaseIsTrue(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Find("THOR").EndOfLine().WithAnyCase(true).Regex()
	if !exp.MatchString("thor") {
		t.Errorf("Expected to match 'thor' with WithAnyCase(true)")
	}
}

func TestShouldMatchMultipleLines(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Anything().Find("Pong").Anything().EndOfLine().SearchOneLine(true).Regex()
	if !exp.MatchString("Ping \n Pong \n Ping") {
		t.Errorf("Expected to match with multiple lines")
	}
}

func TestShouldMatchEmailAddress(t *testing.T) {
	exp := verbalexpressions.NewVerEx().StartOfLine().Word().Then("@").Word().Then(".").Word().EndOfLine().Regex()
	if !exp.MatchString("mail@mail.com") {
		t.Errorf("Expected to match 'mail@mail.com'")
	}
}

func TestShouldMatchURL(t *testing.T) {
	exp := verbalexpressions.NewVerEx().
		StartOfLine().Then("http").Maybe("s").
		Then("://").Maybe("www.").Word().Then(".").Word().Maybe("/").EndOfLine().Regex()
	if !exp.MatchString("https://www.google.com/") {
		t.Errorf("Expected to match 'https://www.google.com/'")
	}
}

File: RepoTransBench/target_projects/Python/Rust/VerbalExpressions_PythonVerbalExpressions/tests/original/verbal_expressions_test.rs
use verbal_expressions::*;
use regex::Regex;

fn assert_not_regex(s: &str, re: &Regex) {
    assert!(!re.is_match(s), "string '{}' should NOT match '{}'", s, re.as_str());
}

#[test]
fn test_should_render_verex_as_string() {
    let mut v = VerEx::new();
    assert_eq!(&*v.add("^$").source(), "^$");
}

#[test]
fn test_should_render_verex_list_as_string() {
    let mut v = VerEx::new();
    let _ = v.add_list(&["^", "[0-9]", "$"]);
    assert_eq!(v.source(), "^[0-9]$");
}

#[test]
fn test_should_match_characters_in_range() {
    let mut v = VerEx::new().start_of_line().range(&["a", "c"]);
    let re = v.regex();
    for character in &["a", "b", "c"] {
        assert!(re.is_match(character));
    }
}

#[test]
fn test_should_not_match_characters_outside_of_range() {
    let mut v = VerEx::new().start_of_line().range(&["a", "c"]);
    let re = v.regex();
    assert_not_regex("d", &re);
}

#[test]
fn test_should_match_characters_in_extended_range() {
    let mut v = VerEx::new().start_of_line().range(&["a", "b", "X", "Z"]);
    let re = v.regex();
    for character in &["a", "b", "X", "Y", "Z"] {
        assert!(re.is_match(character));
    }
}

#[test]
fn test_should_not_match_characters_outside_of_extended_range() {
    let mut v = VerEx::new().start_of_line().range(&["a", "b", "X", "Z"]);
    let re = v.regex();
    assert_not_regex("c", &re);
    assert_not_regex("W", &re);
}

#[test]
fn test_should_match_start_of_line() {
    let mut v = VerEx::new().start_of_line();
    let re = v.regex();
    assert!(re.is_match("text  "), "Not started :(");
}

#[test]
fn test_should_match_end_of_line() {
    let mut v = VerEx::new().start_of_line().end_of_line();
    let re = v.regex();
    assert!(re.is_match(""), "It's not the end!");
}

#[test]
fn test_should_match_anything() {
    let mut v = VerEx::new().start_of_line().anything().end_of_line();
    let re = v.regex();
    assert!(re.is_match("!@#$%¨&*()__+{}"), "Not so anything...");
}

#[test]
fn test_should_match_anything_but_specified_element_when_element_is_not_found() {
    let mut v = VerEx::new().start_of_line().anything_but("X").end_of_line();
    let re = v.regex();
    assert!(re.is_match("Y Files"), "Found the X!");
}

#[test]
fn test_should_not_match_anything_but_specified_element_when_specified_element_is_found() {
    let mut v = VerEx::new().start_of_line().anything_but("X").end_of_line();
    let re = v.regex();
    assert!(!re.is_match("VerEX"), "Didn't found the X :(");
}

#[test]
fn test_should_find_element() {
    let mut v = VerEx::new().start_of_line().find("Wally").end_of_line();
    let re = v.regex();
    assert!(re.is_match("Wally"), "404! Wally not Found!");
}

#[test]
fn test_should_not_find_missing_element() {
    let mut v = VerEx::new().start_of_line().find("Wally").end_of_line();
    let re = v.regex();
    assert!(!re.is_match("Wall-e"), "DAFUQ is Wall-e?");
}

#[test]
fn test_should_match_when_maybe_element_is_present() {
    let mut v = VerEx::new().start_of_line().find("Python2.").maybe("7").end_of_line();
    let re = v.regex();
    assert!(re.is_match("Python2.7"), "Version doesn't match!");
}

#[test]
fn test_should_match_when_maybe_element_is_missing() {
    let mut v = VerEx::new().start_of_line().find("Python2.").maybe("7").end_of_line();
    let re = v.regex();
    assert!(re.is_match("Python2."), "Version doesn't match!");
}

#[test]
fn test_should_match_on_any_when_element_is_found() {
    let mut v = VerEx::new().start_of_line().any("Q").anything().end_of_line();
    let re = v.regex();
    assert!(re.is_match("Query"), "No match found!");
}

#[test]
fn test_should_not_match_on_any_when_element_is_not_found() {
    let mut v = VerEx::new().start_of_line().any("Q").anything().end_of_line();
    let re = v.regex();
    assert!(!re.is_match("W"), "I've found it!");
}

#[test]
fn test_should_match_when_line_break_present() {
    let mut v = VerEx::new().start_of_line().anything().line_break().anything().end_of_line();
    let re = v.regex();
    assert!(re.is_match("Marco \n Polo"), "Give me a break!!");
}

#[test]
fn test_should_match_when_line_break_and_carriage_return_present() {
    let mut v = VerEx::new().start_of_line().anything().line_break().anything().end_of_line();
    let re = v.regex();
    assert!(re.is_match("Marco \r\n Polo"), "Give me a break!!");
}

#[test]
fn test_should_not_match_when_line_break_is_missing() {
    let mut v = VerEx::new().start_of_line().anything().line_break().anything().end_of_line();
    let re = v.regex();
    assert!(!re.is_match("Marco Polo"), "There's a break here!");
}

#[test]
fn test_should_match_when_tab_present() {
    let mut v = VerEx::new().start_of_line().anything().tab().end_of_line();
    let re = v.regex();
    assert!(re.is_match("One tab only\t"), "No tab here!");
}

#[test]
fn test_should_not_match_when_tab_is_missing() {
    let mut v = VerEx::new().start_of_line().anything().tab().end_of_line();
    let re = v.regex();
    assert!(!re.is_match("No tab here"), "There's a tab here!");
}

#[test]
fn test_should_match_when_word_present() {
    let mut v = VerEx::new().start_of_line().anything().word().end_of_line();
    let re = v.regex();
    assert!(re.is_match("Oneword"), "Not just a word!");
}

#[test]
fn test_not_match_when_two_words_are_present_instead_of_one() {
    // Actually tests .tab()! equates to a tab, so:
    let mut v = VerEx::new().start_of_line().anything().tab().end_of_line();
    let re = v.regex();
    assert!(!re.is_match("Two words"), "I've found two of them");
}

#[test]
fn test_should_match_when_or_condition_fulfilled() {
    let mut v = VerEx::new().start_of_line().anything().find("G").OR().find("h").end_of_line();
    let re = v.regex();
    assert!(re.is_match("Github"), "Octocat not found");
}

#[test]
fn test_should_not_match_when_or_condition_not_fulfilled() {
    let mut v = VerEx::new().start_of_line().anything().find("G").OR().find("h").end_of_line();
    let re = v.regex();
    assert!(!re.is_match("Bitbucket"), "Bucket not found");
}

#[test]
fn test_should_match_on_upper_case_when_lower_case_is_given_and_any_case_is_true() {
    let mut v = VerEx::new().start_of_line().find("THOR").end_of_line().with_any_case(true);
    let re = v.regex();
    assert!(re.is_match("thor"), "Upper case Thor, please!");
}

#[test]
fn test_should_match_multiple_lines() {
    let mut v = VerEx::new().start_of_line().anything().find("Pong").anything().end_of_line().search_one_line(true);
    let re = v.regex();
    assert!(re.is_match("Ping \n Pong \n Ping"), "Pong didn't answer");
}

#[test]
fn test_should_match_email_address() {
    let mut v = VerEx::new().start_of_line().word().then("@").word().then(".").word().end_of_line();
    let re = v.regex();
    assert!(re.is_match("mail@mail.com"), "Not a valid email");
}

#[test]
fn test_should_match_url() {
    let mut v = VerEx::new().start_of_line().then("http").maybe("s").then("://").maybe("www.").word().then(".").word().maybe("/").end_of_line();
    let re = v.regex();
    assert!(re.is_match("https://www.google.com/"), "Not a valid email");
}